core/muxing: Have functions on `StreamMuxer` take `Pin<&mut Self>`

thomaseizinger opened the pull request:
# Description

This is the idiomatic way of doing `poll` interface in Rust.



## Links to any relevant issues

<!-- Reference any related issues.-->

https://github.com/libp2p/rust-libp2p/issues/2722
Depends on https://github.com/libp2p/rust-libp2p/pull/2763.
Depends on https://github.com/libp2p/rust-libp2p/pull/2764.
Depends on https://github.com/libp2p/rust-libp2p/pull/2775.
Depends on https://github.com/libp2p/rust-libp2p/pull/2776.

## Open Questions

<!-- Unresolved questions, if any. -->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Introduce `StreamMuxerExt` trait

thomaseizinger committed:
Have functions on `StreamMuxer` use `Pin<&mut Self>`

This is the idiomatic way of defining `poll`-based interfaces in
Rust despite introducing a bit of inconvenience.

To work around this inconvenience, we also introduce a `StreamMuxerExt`
trait allows for better integration with async-await style code and
adds a `_unpin` variant to for each `poll` function of the `StreamMuxer`.

thomaseizinger committed:
Remove `Mutex` from `libp2p-yamux` implementation

Now that the `StreamMuxer` interface has `Pin<&mut Self>` receivers,
we can drop the `Mutex` from the yamux implementation.

elenaf9 commented on core/CHANGELOG.md:
```diff
@@ -2,8 +2,10 @@
+- Use `Pin<&mut Self>` as the receiver type for all `StreamMuxer` poll functions. See [PR XXXX].
 
 [PR 2724]: https://github.com/libp2p/rust-libp2p/pull/2724
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```
```suggestion
- Use `Pin<&mut Self>` as the receiver type for all `StreamMuxer` poll functions. See [PR 2765].

[PR 2724]: https://github.com/libp2p/rust-libp2p/pull/2724
[PR 2765]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```

elenaf9 commented on core/src/muxing.rs:
```diff
@@ -93,5 +104,132 @@ pub trait StreamMuxer {
+}
+
+/// Extension trait for [`StreamMuxer`].
+pub trait StreamMuxerExt: StreamMuxer + Sized {
```
Nice!

thomaseizinger replied:
Glad you like it! I wasn't sure if it isn't over-engineered :)

elenaf9 commented on swarm/src/connection.rs:
```diff
@@ -131,10 +131,10 @@ where
         (
             self.handler.into_connection_handler(),
-            poll_fn(move |cx| self.muxing.poll_close(cx)),
+            poll_fn(move |cx| self.muxing.poll_close_unpin(cx)),
```
```suggestion
            self.muxing.close(),
```
Can we use the `StreamExt::close` here?

thomaseizinger replied:
I tried but unfortunately, we are moving `self` here and `close` doesn't move `self`. It could be made to move it but I am not sure if that isn't too restrictive?

I guess there isn't really a use for accessing the muxer after closing so we can move it into the future.

thomaseizinger replied:
I've changed `StreamMuxerExt::close` to consume `self` which allows us to use it here.

thomaseizinger committed:
Update core/CHANGELOG.md

Co-authored-by: Elena Frank <elena.frank@protonmail.com>

thomaseizinger committed:
Have `StreamMuxerExt::close` consume `self`

The muxer is useless after polling `poll_close` to completion, thus
we can consume `self` in the extension trait method which allows us
to use the function within `libp2p_swarm::Connection`.

thomaseizinger committed:
Merge branch 'master' into feature/pin-mut-self-muxer

thomaseizinger committed:
Fix compile errors and warnings

mxinden commented on core/src/muxing.rs:
```diff
@@ -93,5 +104,132 @@ pub trait StreamMuxer {
+}
+
+/// Extension trait for [`StreamMuxer`].
+pub trait StreamMuxerExt: StreamMuxer + Sized {
```
Fancy

mxinden commented on core/src/muxing.rs:
```diff
@@ -93,5 +104,132 @@ pub trait StreamMuxer {
+where
+    S: StreamMuxer,
+{
+    fn poll_inbound_unpin(
```
Why not provide this as a default implementation on the trait definition itself?

thomaseizinger replied:
I guess we could because we own the trait?

Default impls can be overridden and these are not really meant too. I think that is the only argument for a separate trait.

mxinden replied:
No strong opinion here. One datapoint is that `StreamExt` implements the methods directly:

https://docs.rs/futures/latest/futures/prelude/stream/trait.StreamExt.html

thomaseizinger replied:
Oh, now I understand what you mean! I thought we are discussing having the extension trait vs putting the functions as defaults on `StreamMuxer`.

With the extension trait, I think I prefer default functions too because it co-locates things and should also result in less LoC.

mxinden replied:
Great. To make sure we are on the same page, with the above in mind, this `trait` `impl` would go away @thomaseizinger?

thomaseizinger replied:
The `impl` block would be there but it will be a one-liner as all the implementations are in the trait definition.

One alternative idea would be to remove `StreamMuxerExt` entirely and add the default functions to `StreamMuxer` directly.

I think I mildly prefer the extension trait though because:

- It makes it impossible for implementations to overwrite them, thus guaranteeing that they are the same for all.
- It makes `StreamMuxer` a more coherent trait and we could draw a dependency boundary between `StreamMuxer` and its extension trait if we wanted to.

mxinden reviewed (commented):
:pray: thanks for the patch using `Pin`.

thomaseizinger committed:
Move fn impl to trait definition

thomaseizinger committed:
Merge branch 'master' into feature/pin-mut-self-muxer

thomaseizinger committed:
Add docs to extension trait functions

thomaseizinger committed:
Fix build error

thomaseizinger commented on core/src/muxing.rs:
```diff
@@ -93,5 +104,105 @@ pub trait StreamMuxer {
+
+impl<'a, S> Future for NextInbound<'a, S>
+where
+    S: StreamMuxer + Unpin,
```
I am not sure about this `Unpin` bound and whether we can make it go away or whether it doesn't actually hurt at all.

mxinden replied:
Played around with this as well. I am not sure how one would do this without the `Unpin` bound.

I think this is similar to `StreamExt::next` which as well requires the inner object to be `Unpin`:

> Note that because next doesn’t take ownership over the stream, the [Stream](https://docs.rs/futures/latest/futures/stream/trait.Stream.html) type must be [Unpin](https://doc.rust-lang.org/nightly/core/marker/trait.Unpin.html). If you want to use next with a [!Unpin](https://doc.rust-lang.org/nightly/core/marker/trait.Unpin.html) stream, you’ll first have to pin the stream. This can be done by boxing the stream using [Box::pin](https://doc.rust-lang.org/nightly/alloc/boxed/struct.Box.html#method.pin) or pinning it to the stack using the pin_mut! macro from the pin_utils crate.

https://docs.rs/futures/latest/futures/prelude/stream/trait.StreamExt.html#method.next

thomaseizinger replied:
Cool, thanks for confirming. I'll leave it as is then!

mxinden commented on muxers/yamux/src/lib.rs:
```diff
@@ -39,21 +38,19 @@ use thiserror::Error;
     }
 }
 
-struct Inner<S> {
```
:tada:

mxinden replied:
To expand on this: I am very happy for the `Mutex` around `Inner` to go away. Thanks @thomaseizinger!

thomaseizinger replied:
Me too 😊

mxinden reviewed (approved):
Wonderful to see this happening! :pray:

mxinden committed:
Merge branch 'master' into feature/pin-mut-self-muxer

melekes reviewed (approved):
👍

elenaf9 commented on muxers/mplex/src/lib.rs:
```diff
@@ -85,25 +85,34 @@ where
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
         self.io
             .lock()
```
Couldn't we now get rid of the `Mutex` in `self.io`?

thomaseizinger replied:
Unfortunately no because we need to share the struct between the substreams and the `StreamMuxer` implementation.
